File: packages/sdk-js/src/integrations/alignment.ts
/**
 * Alignment helper — POST a chain-of-thought + proposed action to
 * the AEGIS gateway's `/api/v1/alignment/check` endpoint, get a
 * verdict, and buffer it so the SDK's auto-instrumentation splices
 * it into the next /check call for the same agent.
 *
 * Mirrors the Python helper API.
 */

import { record } from './alignment-state.js';

export type AlignmentProvider = 'anthropic' | 'openai' | 'gemini';
export const PROVIDERS: readonly AlignmentProvider[] = [
  'anthropic',
  'openai',
  'gemini',
] as const;

export interface ProposedAction {
  tool_name: string;
  arguments?: Record<string, unknown>;
}

export interface AlignmentVerdict {
  score: number;
  drifted?: boolean;
  signals?: string[];
  reason?: string;
  model?: string;
  [k: string]: unknown;
}

export interface CheckOptions {
  agentId: string;
  declaredGoal: string;
  proposedAction: ProposedAction;
  thoughtChain?: readonly string[];
  gatewayUrl?: string;
  apiKey?: string;
  provider?: AlignmentProvider;
  model?: string;
  timeoutMs?: number;
  /** If false, don't buffer for the next /check call. Default true. */
  recordForCheck?: boolean;
}

export async function check(opts: CheckOptions): Promise<AlignmentVerdict> {
  const {
    agentId,
    declaredGoal,
    proposedAction,
    thoughtChain,
    gatewayUrl = 'http://localhost:8080',
    apiKey,
    provider,
    model,
    timeoutMs = 10_000,
    recordForCheck = true,
  } = opts;

  if (!agentId) throw new Error('agentId is required');
  if (!declaredGoal) throw new Error('declaredGoal is required');
  if (!proposedAction || !proposedAction.tool_name) {
    throw new Error('proposedAction must include a tool_name');
  }
  if (provider && !PROVIDERS.includes(provider)) {
    throw new Error(`provider must be one of ${PROVIDERS.join(', ')}`);
  }

  const body: Record<string, unknown> = {
    agent_id: agentId,
    declared_goal: declaredGoal,
    thought_chain: Array.isArray(thoughtChain) ? [...thoughtChain] : [],
    proposed_action: {
      tool_name: proposedAction.tool_name,
      arguments: proposedAction.arguments ?? {},
    },
  };
  if (provider) body.provider = provider;
  if (model) body.model = model;

  const headers: Record<string, string> = { 'Content-Type': 'application/json' };
  const key = apiKey ?? (typeof process !== 'undefined' ? process.env?.AEGIS_API_KEY : undefined);
  if (key) headers['X-API-Key'] = key;

  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);
  try {
    const res = await fetch(`${gatewayUrl.replace(/\/$/, '')}/api/v1/alignment/check`, {
      method: 'POST',
      headers,
      body: JSON.stringify(body),
      signal: controller.signal,
    });
    if (!res.ok) {
      const text = await res.text().catch(() => '');
      throw new Error(
        `alignment check failed: HTTP ${res.status}${text ? ` — ${text.slice(0, 200)}` : ''}`,
      );
    }
    const verdict = (await res.json()) as AlignmentVerdict;
    if (recordForCheck) {
      record(agentId, verdict as unknown as Record<string, unknown>);
    }
    return verdict;
  } finally {
    clearTimeout(timer);
  }
}

export { consume as consumeBuffer } from './alignment-state.js';

File: packages/sdk-python/agentguard/integrations/alignment.py
"""
Alignment helper — call ``/api/v1/alignment/check`` from any agent
loop and, optionally, have the verdict auto-flow into the next
``/check`` so the gateway's Policy DSL can react on the same hop.

This is the framework-agnostic entry point. If your agent runs on
LangChain or CrewAI, prefer the dedicated callbacks
(``integrations.langchain.AlignmentCallback``,
``integrations.crewai.make_alignment_step_callback``) — they
capture the chain-of-thought + tool action for free. Reach for
this helper when:

- You're on a framework we don't ship a callback for (autogen,
  raw OpenAI / Anthropic tool loops, in-house agents).
- You want full control over what counts as "the goal" or "the
  thought chain" for a particular hop.

Quick start::

    from agentguard.integrations.alignment import check

    verdict = check(
        agent_id="my-agent",
        declared_goal="Summarise the customer-feedback survey.",
        thought_chain=[
            "Thought: I need to fetch the survey data first.",
        ],
        proposed_action={
            "tool_name": "execute_sql",
            "arguments": {"query": "SELECT * FROM survey"},
        },
        gateway_url="http://localhost:8080",
    )

    if verdict.get("drifted"):
        # The auditor thinks this tool call doesn't serve the
        # declared goal. Decide locally what to do.
        ...

    # The verdict is also buffered in-process keyed by agent_id;
    # the next /check call for the same agent picks it up under
    # `alignment.*` automatically.
"""

from __future__ import annotations

import json
import os
from typing import Any, Optional, Sequence

try:
    import httpx
except ImportError as e:  # pragma: no cover
    raise ImportError(
        "alignment helper requires httpx (a core SDK dep; reinstall agentguard-aegis)."
    ) from e

from . import _alignment_state

PROVIDERS = ("anthropic", "openai", "gemini")


def check(
    *,
    agent_id: str,
    declared_goal: str,
    proposed_action: dict[str, Any],
    thought_chain: Optional[Sequence[str]] = None,
    gateway_url: str = "http://localhost:8080",
    api_key: Optional[str] = None,
    provider: Optional[str] = None,
    model: Optional[str] = None,
    timeout_s: float = 10.0,
    record_for_check: bool = True,
) -> dict[str, Any]:
    """POST a chain-of-thought + proposed action to the AEGIS
    alignment auditor.

    Returns the parsed JSON verdict from the gateway, typically
    ``{"score": float, "drifted": bool, "signals": [str, ...],
       "reason": str, "model": str}``.

    Network or auditor errors are surfaced as ``httpx`` exceptions.
    Callers wanting a fail-open posture should catch
    ``httpx.HTTPError`` and treat it as "no verdict."

    Parameters
    ----------
    agent_id
        Stable identifier for the calling agent. Used as the buffer
        key for the closed-loop bridge.
    declared_goal
        Plain-English statement of what the current agent run is
        trying to accomplish. The auditor uses this as the anchor.
    proposed_action
        ``{"tool_name": ..., "arguments": {...}}`` — the tool the
        agent is about to dispatch.
    thought_chain
        Optional list of intermediate reasoning steps in order. If
        omitted, the auditor scores from goal + action alone.
    provider, model
        Optional overrides for the upstream LLM judge. Defaults
        come from the gateway's configured provider.
    record_for_check
        When True (default), the verdict is stashed in the
        closed-loop buffer so the SDK's auto-instrumentation will
        splice it into the next ``/check`` payload for this
        ``agent_id``.
    """
    if not agent_id:
        raise ValueError("agent_id is required")
    if not declared_goal:
        raise ValueError("declared_goal is required")
    if not isinstance(proposed_action, dict) or "tool_name" not in proposed_action:
        raise ValueError("proposed_action must be a dict with at least a tool_name")
    if provider is not None and provider not in PROVIDERS:
        raise ValueError(
            f"provider must be one of {PROVIDERS}; got {provider!r}"
        )

    body: dict[str, Any] = {
        "agent_id": agent_id,
        "declared_goal": declared_goal,
        "thought_chain": list(thought_chain or []),
        "proposed_action": {
            "tool_name": proposed_action["tool_name"],
            "arguments": dict(proposed_action.get("arguments") or {}),
        },
    }
    if provider:
        body["provider"] = provider
    if model:
        body["model"] = model

    headers = {"content-type": "application/json"}
    key = api_key or os.environ.get("AEGIS_API_KEY", "")
    if key:
        headers["x-api-key"] = key

    url = gateway_url.rstrip("/") + "/api/v1/alignment/check"
    with httpx.Client(timeout=timeout_s) as client:
        resp = client.post(url, headers=headers, content=json.dumps(body))
    resp.raise_for_status()
    verdict = resp.json()

    if record_for_check and isinstance(verdict, dict):
        _alignment_state.record(agent_id, verdict)

    return verdict


def consume(agent_id: str) -> Optional[dict[str, Any]]:
    """Pop the most recent verdict for ``agent_id`` if still
    fresh (≤ 30 s old). Intended for callers that want to inspect
    the verdict before invoking the next tool, instead of relying
    on the auto-instrumentation's silent splice."""
    return _alignment_state.consume(agent_id)
